src/libPMacc/include/eventSystem/events/EventPool.hpp

```diff
@@ -23,79 +23,119 @@
 
 #pragma once
 
-#include "eventSystem/events/CudaEvent.hpp"
+#include "eventSystem/events/CudaEvent.def"
+#include "eventSystem/events/CudaEventHandle.hpp"
 #include "debug/VerboseLog.hpp"
 #include "pmacc_types.hpp"
 
 #include <vector>
+#include <list>
+#include <stdexcept>
 
 namespace PMacc
 {
 
-    /**
-     * Manages a pool of cudaEvent_t objects and gives access to them.
-     */
+    /** Manages a pool of cudaEvent_t objects and gives access to them. */
     class EventPool
     {
     public:
 
-        /**
-         * Constructor.
-         * adds a single cuda event to the pool
+        /** Returns a free cuda event
+         *
+         * @return free cuda event
          */
-        EventPool()
+        CudaEventHandle pop( )
         {
-            addEvents(1);
-            currentEventIndex = 0;
+            if( freeEvents.size( ) != 0 )
+            {
+                CudaEventHandle result = freeEvents.front( );
+                freeEvents.pop_front( );
+                return result;
+            }
+            createEvents( );
+            return pop( );
         }
 
-        /**
-         * Destructor.
-         * destroys all cuda events in the pool
+
+        /** add CudaEvent to the pool
+         *
+         * the pool takes the ownership of the pointer
+         *
+         * @param ev pointer to CudaEvent
          */
-        virtual ~EventPool()
+        void push( CudaEvent* const ev )
         {
-            for (size_t i = 0; i < events.size(); i++)
+            /* Guard that no event is added during the pool is closed (shutdown phase).
+             * This function is also called during the destructor is evaluated.
```

Comment:
Grammar: `This function is also called during the evaluation of the destructor.`